src/main/java/sg/ncl/DataController.java

```diff
@@ -0,0 +1,236 @@
+
+    private static final String CONTRIBUTE_DATA_PAGE = "data_contribute";
+
+    @RequestMapping
+    public String data(Model model, HttpSession session) throws Exception {
+        DatasetManager datasetManager = new DatasetManager();
+
+        HttpEntity<String> request = createHttpEntityHeaderOnly();
+        ResponseEntity response = restTemplate.exchange(properties.getData(), HttpMethod.GET, request, String.class);
+        String dataResponseBody = response.getBody().toString();
+
+        JSONArray dataJsonArray = new JSONArray(dataResponseBody);
+        for (int i = 0; i < dataJsonArray.length(); i++) {
+            JSONObject dataInfoObject = dataJsonArray.getJSONObject(i);
+            Dataset dataset = extractDataInfo(dataInfoObject.toString());
+            datasetManager.addDataset(dataset);
+        }
+
+        model.addAttribute("allDataMap", datasetManager.getDatasetMap());
+        return "data";
+    }
+
+    @RequestMapping(value={"/contribute", "/contribute/{id}"}, method= RequestMethod.GET)
+    public String contributeData(Model model, @PathVariable Optional<String> id) throws Exception {
+        if (id.isPresent()) {
+            HttpEntity<String> request = createHttpEntityHeaderOnly();
+            ResponseEntity response = restTemplate.exchange(properties.getDataset(id.get()), HttpMethod.GET, request, String.class);
+            String dataResponseBody = response.getBody().toString();
+            JSONObject dataInfoObject = new JSONObject(dataResponseBody);
+            Dataset dataset = extractDataInfo(dataInfoObject.toString());
+            model.addAttribute("dataset", dataset);
+        } else {
+            model.addAttribute("dataset", new Dataset());
+        }
+        return CONTRIBUTE_DATA_PAGE;
+    }
+
+    @RequestMapping(value={"/contribute", "/contribute/{id}"}, method=RequestMethod.POST)
+    public String validateContributeData(@Valid @ModelAttribute("dataset") Dataset dataset,
+                                         BindingResult bindingResult,
+                                         Model model, @PathVariable Optional<String> id,
+                                         HttpSession session) throws WebServiceRuntimeException {
+        if (bindingResult.hasErrors()) {
+            StringBuilder message = new StringBuilder();
+            message.append("Error(s):");
+            message.append("<ul class=\"fa-ul\">");
+            for (ObjectError objectError : bindingResult.getAllErrors()) {
+                FieldError fieldError = (FieldError) objectError;
+                message.append("<li><i class=\"fa fa-exclamation-circle\"></i> ");
+                message.append(fieldError.getField());
+                message.append(" ");
+                message.append(fieldError.getDefaultMessage());
+                message.append("</li>");
+            }
+            message.append("</ul>");
+            model.addAttribute("message", message.toString());
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "message" 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1192)